src/upgrade/migrator.js

```diff
@@ -0,0 +1,88 @@
+     */
+    constructor(from_client, to_client, collection_list, batch_size) {
+        this.collection_list = collection_list;
+        this.from_client = from_client;
+        this.to_client = to_client;
+        this.batch_size = batch_size || 100;
+    }
+
+    async migrate_db() {
+        this.to_client.define_collection({ name: 'migrate_status' });
+        for (const collection of this.collection_list) {
+            await this.from_client.define_collection(_.omit(collection, 'db_indexes'));
+            await this.to_client.define_collection(collection);
+        }
+        await this.from_client.connect(true);
+        await this.to_client.connect();
+        this.upgrade_table = this.to_client.collection('migrate_status');
+        this.migrate_status = await this.upgrade_table.findOne({});
+        if (!this.migrate_status) {
+            this.migrate_status = {
+                _id: this.to_client.new_object_id(),
+                collection_index: 0,
+                marker: undefined,
+            };
+            await this.upgrade_table.insertOne(this.migrate_status);
+        }
+        const start_index = this.migrate_status.collection_index;
+        for (let i = start_index; i < this.collection_list.length; ++i) {
+            this.migrate_status.collection_index = i;
+            await P.retry({ attemps: 3, delay_ms: 10000, func: async () => this._migrate_collection(i)});
+            this.migrate_status.marker = undefined;
+            await this.upgrade_table.updateOne({}, { $set: { collection_index: i + 1 }, $unset: { marker: 1}});
+        }
+        await this.from_client.disconnect();
+        await this.to_client.disconnect();
+    }
+
+    async _migrate_collection(collection_index) {
+        const col_name = this.collection_list[collection_index].name;
+        console.log(`migrating ${col_name}`);
+        const from_col = this.from_client.collection(col_name);
+        const to_col = this.to_client.collection(col_name);
+        let done = false;
+        let marker = this.migrate_status.marker;
+        let total = 0;
+        while (!done) {
+            const docs = await from_col.find({ _id: marker ? { $gt: marker } : undefined }, { limit: this.batch_size, sort: { _id: 1 }});
+            if (docs.length > 0) {
+                try {
+                    await to_col.insertMany(docs, {
+                        ordered: false
+                    });
+                } catch (err) { // if duplicate key - continue
+                    if (err.code !== '23505') throw err;
+                }
+                total += docs.length;
+                console.log(`migrated ${total} documents to table ${col_name}`);
+                marker = docs[docs.length - 1]._id;
+                await this.upgrade_table.updateOne({}, { $set: { marker }});
+            } else {
+                done = true;
+            }
+        }
+        console.log(`completed migration of ${total} documents to table ${col_name}`);
```

Comment:
as a sanity check, maybe check that the sizes of the collections are the same

Reply:
done